xds/src/main/java/com/linecorp/armeria/xds/client/endpoint/EndpointsPool.java

```diff
@@ -48,36 +49,47 @@ final class EndpointsPool implements AsyncCloseable {
     }
 
     void updateClusterSnapshot(ClusterSnapshot newSnapshot, Consumer<List<Endpoint>> endpointsListener) {
-        // clean up the old endpoint and listener
+        // it is very important that the listener is removed first so that endpoints aren't deemed
+        // unhealthy due to closing a HealthCheckedEndpointGroup
         delegate.removeListener(listener);
         delegate.closeAsync();
 
         // set the new endpoint and listener
-        delegate = XdsEndpointUtil.convertEndpointGroup(newSnapshot);
+        delegate = XdsEndpointUtil.convertEndpointGroup(newSnapshot, prevAttrs);
         listener = endpoints -> eventExecutor.execute(
-                () -> endpointsListener.accept(attachTimestampsAndDelegate(endpoints)));
+                () -> endpointsListener.accept(cacheAttributesAndDelegate(endpoints)));
         delegate.addListener(listener, true);
     }
 
-    private List<Endpoint> attachTimestampsAndDelegate(List<Endpoint> endpoints) {
+    private List<Endpoint> cacheAttributesAndDelegate(List<Endpoint> endpoints) {
         final long defaultTimestamp = System.nanoTime();
-        final ImmutableMap.Builder<Endpoint, Long> timestampsBuilder = ImmutableMap.builder();
         final ImmutableList.Builder<Endpoint> endpointsBuilder = ImmutableList.builder();
+        final ImmutableMap.Builder<Endpoint, Attributes> prevAttrsBuilder = ImmutableMap.builder();
         for (Endpoint endpoint: endpoints) {
-            final long timestamp;
-            if (hasCreatedAtNanos(endpoint)) {
-                timestamp = createdAtNanos(endpoint);
-            } else {
-                timestamp = createdTimestamps.getOrDefault(endpoint, defaultTimestamp);
-            }
-            final Endpoint endpointWithTimestamp = withCreatedAtNanos(endpoint, timestamp);
-            timestampsBuilder.put(endpointWithTimestamp, timestamp);
+            final long timestamp = computeTimestamp(endpoint, defaultTimestamp);
+            final Endpoint endpointWithTimestamp = endpoint.withAttr(CREATED_AT_NANOS_KEY, timestamp);
             endpointsBuilder.add(endpointWithTimestamp);
+            prevAttrsBuilder.put(endpoint, endpointWithTimestamp.attrs());
         }
-        createdTimestamps = timestampsBuilder.buildKeepingLast();
+        prevAttrs = prevAttrsBuilder.buildKeepingLast();
         return endpointsBuilder.build();
     }
 
+    private long computeTimestamp(Endpoint endpoint, long defaultTimestamp) {
+        if (hasCreatedAtNanos(endpoint)) {
+            return createdAtNanos(endpoint);
```

Comment:
I think we don't have to create another endpoint when the endpoint has the value:
```java
final Endpoint endpointWithTimestamp = withTimestamp(endpoint, defaultTimestamp);
...
private Endpoint withTimestamp(Endpoint endpoint, long defaultTimestamp) {
    if (hasCreatedAtNanos(endpoint)) {
        return endpoint;
    }
    Long timestamp = null;
    final Attributes prevAttr = prevAttrs.get(endpoint);
    if (prevAttr != null) {
        timestamp = prevAttr.attr(CREATED_AT_NANOS_KEY);
    }
    return endpoint.withAttr(CREATED_AT_NANOS_KEY, timestamp != null ? timestamp : defaultTimestamp);
}
```